pkg/blockbuilder/blockbuilder.go

```diff
@@ -427,39 +475,73 @@ func (b *BlockBuilder) blockUploaderForUser(ctx context.Context, userID string)
 	}
 }
 
-func (b *BlockBuilder) finalizePartition(ctx context.Context, uncommittedRec, checkpointRec *kgo.Record) error {
-	// If there is an uncommitted record, rewind the client to the record's offset to consume it on the next cycle.
-	if uncommittedRec != nil {
-		part := uncommittedRec.Partition
-		b.kafkaClient.SetOffsets(map[string]map[int32]kgo.EpochOffset{
-			b.cfg.Kafka.Topic: {
-				part: {
-					Epoch:  uncommittedRec.LeaderEpoch,
-					Offset: uncommittedRec.Offset - 1,
-				},
-			},
-		})
+func (b *BlockBuilder) finalizePartition(ctx context.Context, commitRec, lastRec *kgo.Record, blockEnd int64) error {
+	if commitRec == nil {
+		return nil
 	}
+	// Rewind the offset to the commit record so that when the partition is read again by this
+	// block builder, it starts at the commit point.
+	b.kafkaClient.SetOffsets(map[string]map[int32]kgo.EpochOffset{
+		b.cfg.Kafka.Topic: {
+			commitRec.Partition: {
+				Epoch:  commitRec.LeaderEpoch,
+				Offset: commitRec.Offset,
+			},
+		},
+	})
 
-	if checkpointRec != nil {
-		// TODO(codesome): persist uncommittedRec with checkpoint's metadata
-		err := b.kafkaClient.CommitRecords(ctx, checkpointRec)
-		if err != nil {
-			// TODO(codesome): add metric
-			return err
-		}
+	var off kadm.Offsets
+	off.Add(kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRec.Offset,
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), lastRec.Timestamp.UnixMilli(), blockEnd),
+	})
+	committed, err := kadm.NewClient(b.kafkaClient).CommitOffsets(ctx, b.cfg.Kafka.ConsumerGroup, off)
+	if err != nil {
+		return err
+	} else if !committed.Ok() {
+		return committed.Error()
 	}
```

Comment:
This still doesn't work when I run the branch against real Kafka. It fails with this one (I've formatted the log line for readability):

```
ts=2024-06-12T10:48:15.35956496Z
  caller=blockbuilder.go:307
  level=error
  msg="failed to consume partition"
  err="UNKNOWN_MEMBER_ID: The coordinator is not aware of this member."
  part=2

ts=2024-06-12T10:53:32.299565461Z
  caller=blockbuilder.go:370
  level=info
  msg="done consuming partition"
  part=2
  dur=23.669061273s
```

When inspecting the group with `kafka-consumer-groups`, it shows that nothing has been committed:

```
[appuser@6fd9e0581032 ~]$ kafka-consumer-groups --bootstrap-server kafka:9092 --group mimir-block-builder --describe

GROUP               TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                          HOST            CLIENT-ID
mimir-block-builder mimir-ingest    9          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    0          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    3          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    1          0               24194           24194           block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    7          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    5          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    6          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    4          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    8          0               0               0               block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
mimir-block-builder mimir-ingest    2          0               24215           24215           block-builder-1-a623dfac-a1a9-4f9b-b000-ec16e48af7b8 /172.20.0.16    block-builder-1
```